proposals/function-pointers.md

```diff
@@ -0,0 +1,245 @@
+have the same signature they have an identity conversion no matter what the name is. 
+
+``` csharp
+unsafe class Example {
+    funcptr int Func1(int left, int right);
+    funcptr int Func2(int x);
+    funcptr int Func3(int x, int y);
+    funcptr cdecl int Func4(int x, int y);
+
+    void Conversions() {
+        Func1 p1 = ...;
+        Func2 p2 = ...;
+
+        Func3 p3 = p1; // okay Func1 and Func3 have compatible signatures
+        Console.WriteLine(p3 == p1); // True
+        Func3 p4 = p2; // error: Func3 and Func2 have incompatible signatures (parameter)
+        Func4 p5 = p1; // error: Func4 and Func1 have incompatible signatures (calling convention)
+    }
+}
+```
+
+In addition to declaring a named `funcptr` type, as you declare a `delegate`, it is possible to use an unnamed 
+`funcptr` type without first declaring it. This type can be used anywhere a type declaration would occur:
+
+``` csharp
+unsafe struct Example {
+    funcptr int (int) Field;
+    unsafe void UnnamedExample(funcptr int(int) ptr) {
+        int x = ptr(42);
+        Field = ptr;
+        ...
+    }
+}
+```
+
+A `funcptr` type is a pointer type which means it has all of the capabilities and restrictions of a standard pointer
+type:
+- Only valid in an `unsafe` context.
+- Methods which contain a `funcptr` parameter or return type can only be called from an `unsafe` context.
+- Cannot be converted to `object`.
+- Cannot be used an a generic argument.
+- Can implicitly convert `funcptr` to `void*`.
+- Can explicitly convert from `void*` to `funcptr`.
+
+Restrictions:
+- Cannot overload when the only difference in parameter types is the name of the function pointer. 
+- Custom attributes cannot be applied to a `funcptr` or any of its elements.
+- A `funcptr` parameter cannot be marked as `params`
+- A `funcptr` type has all of the restrictions of a normal pointer type.
+
+### Allow addresss-of to target methods
+
+Method groups will now be allowed as arguments to an address-of expression. The type of such an 
+expression will be an unnamed `funcptr` which has the equivalent signature of the target method and a managed 
+calling convention:
+
+``` csharp
+unsafe class Util { 
+    public static void Log() { } 
+
+    funcptr void Action();
+    funcptr int Func();
+    void Use() {
+        funcptr void() ptr1 = &Util.Log; 
+        Action ptr2 = &Util.Log;
+
+        // Error: type "funcptr void()" not compatible with "funcptr int()";
+        Func ptr3 = &Util.Log; 
+
+        // Okay. Conversion to void* is always allowed.
+        void* v = &Util.Log;
+   }
+}
+```
+
+The conversion of an address-of method group to `funcptr` has roughly the same process as method group to `delegate`  
+conversion. There are two additional restrictions to the existing process:
+- Only members of the method group that are marked as `static` will be considered. 
+- Only a `funcptr` with a managed calling convention can be the target of such a conversion.
```

Comment:
This makes it unusable for interop that is one of the main scenarios for this.

Can we tie this with the `NativeCallableAttribute` to make it work for interop? 
- If the target method is not marked with `NativeCallableAttribute`, the result is function pointer with the managed calling convention
- If the target method is marked with `NativeCallableAttribute`, the result is function pointer with the unmanaged calling convention specified in the `NativeCallableAttribute`.

Reply:
That is pretty doable. We'd likely need to expose that attribute though and make it a supported API

Reply:
Is NativeCallableAttribute exposed in .NET Framework? This will make this feature not useful on .NET Framework or until we decide to put NativeCallableAttribute in .NET Framework. Any reason why we can't have this for all supported calling conventions with types already present on desktop and .net core?

Reply:
> Is NativeCallableAttribute exposed in .NET Framework

Nope. (In fact, it is not exposed in the public surface of any shipping runtime today.) 

> Any reason why we can't have this for all supported calling conventions with types already present on desktop and .net core?

What would you propose? I do not think that the runtime support required to make this work well exist in desktop.